hpy/devel/include/hpy/runtime/argparse.h

```diff
@@ -3,10 +3,10 @@
 
 #include "hpy.h"
 
-HPyAPI_RUNTIME_FUNC(int)
+_HPy_HIDDEN int
```

Comment:
I'm a bit uncomfortable just labeling these `_HPy_HIDDEN` because it doesn't explain why these functions are marked hidden. What differentiates them from `HPyAPI_IMPL`?  I don't know what the name should be, buI feel we *should* have a name for them so it's clear what sort of thing they are.

Reply:
They are marked `_HPy_HIDDEN` because it's a good practice, for the very same reason why you mark most other functions as `static`, i.e. because you want export nothing by default unless you explicitly export the symbol, and C has the wrong default :man_shrugging: .
From this point of view it has nothing to do with the fact that they are or are not "API functions". E.g. we do the very same for `create_method_defs`:
https://github.com/hpyproject/hpy/blob/9dd30240456abe5b1d9eb489dd0d2f9a5784c9a4/hpy/devel/src/runtime/ctx_type.c#L90-L93

That said, it might be a good idea to mark them somehow, to be self-documenting. What about `HPyAPI_HELPER` ?

Reply:
I like `HPyAPI_HELPER`, although I think there are really two kinds of functions here:
* `HPyAPI_HELPER`s like `HPyArg_Parse`
* `HPyAPI_INTERNAL`s like `ctx_...`
with the difference being that helpers are intended to be used by writers of extensions, but never exported as symbols and internal functions are neither.

Reply:
yes, I'm proposing to have:
1. `HPyAPI_FUNC`
2. `HPyAPI_IMPL` (what you call `HPyAPI_INTERNAL)
3. `HPyAPI_HELPER`

I like `IMPL` better than `INTERNAL` because "internal" sounds like an internal API which is not meant to be exposed to the user, i.e. things like CPython's  `_PyTuple_FromArray` which nowadays are in `Include/internal/*.h`.
Out `ctx` functions are different... they are not internal APIs, they are the implementation of user-facing APIs (but you need to call them through the `HPy_*` functions because you do different things depending on whether you are using the CPython or Universal ABI).

I agree that maybe we should write some internal doc to explain all of this

Reply:
I introduced `HPyAPI_HELPER` and a big explanatory comment in 195be06

Reply:
I like the FUNC / IMPL / HELPER naming and woot on the big comment! :D